### DISCUSSION TITLE
Using Gain and Bias to match images, specifically exposure

### DISCUSSION THREAD

**Discussion by memphis-10** [ImageMagick/ImageMagick#5467] [Digital Image Processing]
Hi,

I have a Canon CR2 file that is slightly underexposed where 18% is averaging about 35% RGB.

When I raise the exposure in the Canon Digital Photo Professional app by 0.8, the grey patch is closer to 50% and looks "correct" (how I recall the scene looking).

*note: the photograph was taken under late afternoon sun, so 18% is not perfect grey, but the variance isn't a concern at the moment.

I would like an Imagemagick method to achieve an identical result - knowing both the input and output file values.  I've gotten close, but need help.

Both files, original and corrected, were exported from the Canon app to TIF as a starting point.

original (resized and converted to jpg for posting)
`magick original.TIF -resize 600x400 original_sm.jpg`
![IMG_6012_canon_sm-0](https://user-images.githubusercontent.com/88346417/185805379-f52c54ff-f034-4457-8604-ad908b7e1db8.jpg)

corrected (resized and converted to jpg for posting)
`magick corrected.TIF -resize 600x400 corrected_sm.jpg`
![IMG_6012_Canon_exp_only_sm](https://user-images.githubusercontent.com/88346417/185805478-1e8b1e8f-cc4c-452a-9ea7-25bee3dfe4a3.jpg)

I am using GIMP to get the mean and standard deviation values, then I plug them into a calculator I made.
The calculation is as follows (thanks to @snibgo): 

gain = (corrected) SD / (original) SD
bias = (corrected) mean - (original) mean * gain

This is a screenshot of the calculator's input and results
![Screen Shot 2022-08-21 at 1 35 32 PM](https://user-images.githubusercontent.com/88346417/185805803-e7944faf-da53-48c8-a6b0-f9828feb5468.png)

This is where I'm not clear:
1) have I made the calculations correctly?
2) how do I implement those values?  

There seems to be a gap between what I know and how to utilize the information correctly.

My closest attempt is close, but still off slightly.  If you toggle between this IM corrected version and the Canon corrected version, there is an obvious difference 

original.TIF converted in IM and saved as jpg for posting:
`magick original.TIF -channel r -function polynomial "1.202,0.0573" -channel g -function polynomial "1.2044, 0.0591" -channel b -function polynomial "1.2326,0.0419" -resize 600x400 IM_corrected_sm.jpg`
![IM_corrected_sm](https://user-images.githubusercontent.com/88346417/185806104-0128de18-6dab-475b-8dbf-3f38b3d0a2bc.jpg)

What am I missing?  Thanks a ton

Side note: if there's a better method, I'm open.  I prefer to avoid clipping but also have limited experience inside IM and the complex math gets away from me.

Thanks

**Comment by fmw42** [comment]
What is your IM version and platform/OS?

**Reply by memphis-10** [reply]
Sorry:

ImageMagick 7.1.0-10 Q16-HDRI x86_64 2021-10-10 
Mac OS

**Comment by fmw42** [comment]
You could use -color-matrix. See https://imagemagick.org/script/command-line-options.php#color-matrix

**Reply by memphis-10** [reply]
Cheers, thank you.  I should add what the overall objective is, which is maybe/ maybe not possible.

I have the same photo taken with a few cameras and film stocks.  However, the Canon CR2 is the only one with known original and corrected values.  

What I want to do is use the Canon image to create an IM string which corrects the original image to match the corrected one, then apply that to the other cameras/film stocks.  In my limited experience, I'm looking for the simplest method which I could then apply across the board vs. specific color matrix and matching for the one image.

For example: raising exposure 1/2 stop - but I want to build the string vs. running a script so I can control/reuse certain portions as needed.

Does that make sense?

**Comment by fmw42** [comment]
You could also do the computations from mean and std to gain and bias inside the command line. Also you could find the mean and std inside the command line also.  See -set option:rmean "%[fx:mean.r] to create it and then to use them use "%[rmean]"

**Reply by memphis-10** [reply]
Is this more accurate than GIMP's mean and std?

**Reply by fmw42** [reply]
They should be equivalent.  try

`magick -precision 10 image -format "%[fx:mean.r],%[fx:mean.g],%[fx:mean.b]" info:`

replace image with your actual image file.suffix  

Change the precision value as desired. The default is 6, I believe

Likewise you can do the same or include the standard_deviation.x where x is r, g, or b. Or given your desired values you can extract those and compute your gain and bias in an %[fx: ...] expression.

**Reply by memphis-10** [reply]
Thank you, that does give the same results as Gimp.  I'm working on getting the values from IM now.

**Comment by snibgo** [comment]
I suppose you have read my [Gain and bias](http://im.snibgo.com/gainbias.htm) page.

I don't know what "raise the exposure in the Canon Digital Photo Professional app by 0.8" does. I suppose it multiplies the three RGB channels by some number, and I suspect it does so *in linear colourspace*, not sRGB. If the sRGB transfer response curve (TRC) was a power curve, this would have same effect as multiplying the channels by some other number in sRGB space.

As the sRGB TRC isn't exactly a power curve, the effect isn't exactly the same. But the difference is small.

If I use `magick gbOrig.jpg -verbose info:` (on your "image A") I get this:
```
    Red:
      min: 0  (0)
      max: 241 (0.945098)
      mean: 69.0308 (0.270709)
      median: 60 (0.235294)
      standard deviation: 50.4575 (0.197873)
      kurtosis: 0.147059
      skewness: 0.831564
      entropy: 0.929875
    Green:
      min: 0  (0)
      max: 231 (0.905882)
      mean: 61.0215 (0.2393)
      median: 55 (0.215686)
      standard deviation: 44.9744 (0.17637)
      kurtosis: 1.28494
      skewness: 1.19996
      entropy: 0.908381
    Blue:
      min: 0  (0)
      max: 225 (0.882353)
      mean: 42.9639 (0.168486)
      median: 27 (0.105882)
      standard deviation: 43.0702 (0.168903)
```
The means are the same as you get from Gimp. But the standard deviations are different. They might calculate SD differently ("population" or "sample" methods), and I suppose you measured the full image where I measure the resized 600x400, and I am working from JPEGs, but will those account for the red channel 0.197873 from IM but 0.203 from Gimp? This is a factor of 1.026, or 2.6%, which will be noticeable, 

The "corrected" image also has the same means but different SDs.

Moving on, I call my script:
```
call %PICTBAT%imgGainBias gbOrig.jpg gbCorr.jpg x.png
```
... and that can tell me the gain and bias numbers:
```
f:\web\im>set igb_gb_gain_R=1.18392967845831576
f:\web\im>set igb_gb_gain_G=1.179139284714595393
f:\web\im>set igb_gb_gain_B=1.212619444231479493

f:\web\im>set igb_gb_bias_R=0.06253223526362766682
f:\web\im>set igb_gb_bias_G=0.06454001597817893288
f:\web\im>set igb_gb_bias_B=0.04419022933327482194
```
As the SDs are different, the gain and bias numbers are also different.

We can compare results. First, compare your IM result to your "corrected" result (ie with Gimp SD numbers):
```
f:\web\im>%IMG7%compare -metric RMSE gbCorr.jpg gbIM.jpg NULL:
3271.15 (0.0499146)
```
Then, compare the result from my script to your "corrected" result (ie with IM SD numbers):
```
f:\web\im>%IMG7%compare -metric RMSE gbCorr.jpg x.png NULL:
1944.96 (0.0296781)
```
The SDs from IM give a result closer to the "corrected" version.

**Reply by memphis-10** [reply]
Thank you for taking the time to write all of this out.

When I run `magick gbOrig.TIF -verbose info:` on the original TIF, I get values that are nearly identical to GIMP, and therefore puts me back at the same starting point.  

I think the small jpg must be throwing the SD off from the original.

    `Red:
          min: 0  (0)
          max: 64895 (0.990234)
          mean: 17753.9 (0.270907)
          median: 10501 (0.160235)
          standard deviation: 13269.8 (0.202484)
          kurtosis: 0.00905954
          skewness: 0.813515
          entropy: 0.951454
        Green:
          min: 0  (0)
          max: 62334 (0.951156)
          mean: 15681.8 (0.239289)
          median: 12614 (0.192477)
          standard deviation: 11866.6 (0.181072)
          kurtosis: 1.06852
          skewness: 1.17231
          entropy: 0.93804
        Blue:
          min: 0  (0)
          max: 57692 (0.880323)
          mean: 11051.2 (0.168631)
          median: 4674 (0.0713207)
          standard deviation: 11272.3 (0.172005)`

**Comment by snibgo** [comment]
> When I run magick gbOrig.TIF -verbose info: on the original TIF, I get values that are nearly identical to GIMP, and therefore puts me back at the same starting point.

Good, excellent.

In your red channel:
```
      max: 64895 (0.990234)
```
... so multiplying values by more than about 1% will cause clipping when images are saved in non-HDRI formats. Perhaps your Canon software does some kind of roll-off in highlights to avoid clipping. See [Putting OOG back in the box](http://im.snibgo.com/oogbox.htm) for some possible methods. If roll-off or clipping applies, this messes with the mean and SD calculations.

"Increasing exposure" by one stop should be equivalent to multiplying by 2 in linear space. Increasing by N stops is equivalent to multiplying by 2^N:
```
  V' = V * 2^N.
```
For example, increasing by 0.8 stops is multiplying in linear space by 2^0.8 = 1.7411.

Converting from linear to non-linear sRGB is approximately raising to the power (1/2.2). Taking a non-linear sRGB image, increasing exposure by N stops, and converting back to non-linear is:
```
V' = ( V^2.2 * 2^N ) ^ (1/2.2)
   = V^2.2^(1/2.2) * 2^N^(1/2.2)
   = V             * 2^(N/2.2)
```
For example, increasing by 0.8 stops is multiplying in non-linear sRGB space by approximately 2^(0.8/2.2) = 1.2866.

We can compare this to your calculated gain numbers (from Gimp SD): 1.2020, 1.2044 and 1.2326. Close, but not very close.

If the Canon software *only* multiplies by some number, than the gain for the three channels should be equal, and the bias for all channels should be zero.

**Reply by memphis-10** [reply]
Ah cheers, this makes sense. Thanks for taking the time.

I would bet that Canon has highlight safety, and I was beginning to assume there would be some sort of issue with their computations that would keep me from getting an exact match with gain and bias in IM. I tried straight multiply early on, but that wasn't as good as gain and bias, so I don't think that's their method.

Ultimately, the end result of the work I'm doing isn't to match these two images, rather I'm trying to obtain a general compensation that could be applied to other images with similar brightness shot under the same conditions.

For now, I'll consider myself very close and will test the values I currently have on the other images when I get them all ready.

**Comment by snibgo** [comment]
Another useful analysis is to find the graph of the clut that would make the histogram of gbCorr.jpg match the histogram of gcOrig.jpg.
```
set mhDEBUG=1
call %PICTBAT%matchHisto gbOrig.jpg gbCorr.jpg x3.png
call %PICTBAT%graphLineCol gbOrig_mhcl.miff
```
The graph has values in gbOrig on the x-axis, and gbCorr on the y-axis.

![gbOrig_mhcl_glc](https://user-images.githubusercontent.com/33812368/185823569-226719d6-fcaf-4ba3-aa8f-ea3a47532e40.png)

The major effect is a multiplier, so input 0.85 becomes output 1.0. Then there is a slight power curve that lifts mid-tones, so input 0.425 becomes output 0.57 instead of 0.50. Then there are some twiddly bits at the top end, possibly to preserve highlights, but also possibly from JPEG artefacts.

**Reply by memphis-10** [reply]
Thank you.  That's a little advanced for me out of the gate, but I'm excited to learn more.  

Are those functions possible in IM alone without a script?  I get a little lost trying to decipher what's BAT and what's IM.

Ultimately, say I'm able to figure out the histogram matching, would I be able to apply the change to multiple images which are similar, or would the shift only work for identical images?

(i.e. could I apply the same shift to a series of images shot under the same exposure and lighting conditions, but different framing or different cameras?)

**Comment by fmw42** [comment]
You might try using a Hald CLUT. You take a HALD image, then process it the way you want. Then use that with -hald-clut to process all other images. This is a global process, not local. So it does the same everywhere.  See https://imagemagick.org/Usage/color_mods/#hald-clut

**Reply by memphis-10** [reply]
Cheers, thank you.  I appreciate that idea; I have worked with CLUTs on a different level, but this would be a great time to try again.

**Reply by memphis-10** [reply]
I just wanted to come back and say thank you, and recognize your histmatch script which I was able to successfully run on a different image, which is part of the overall larger processing I'm having to do.  I will certainly try it in this example next.  Thank you.

**Reply by memphis-10** [reply]
@fmw42 I have a question about some specific results of your histmatch script.  Is this a good place to ask, or should I take it to another thread?  thank you.

**Reply by fmw42** [reply]
Post to https://github.com/ImageMagick/Freds-Scripts/discussions

**Comment by snibgo** [comment]
> Are those functions possible in IM alone without a script? I get a little lost trying to decipher what's BAT and what's IM.

"call" is a Windows command that executes a .BAT script file. I publish these on my website im.snibgo.com . Most of my scripts can be easily translated to bash or any other shell. However, `matchHisto.bat` calls a process module. These are like plug-ins for IM, written in C, and compiled into the IM build. I publish these, mostly at [Process modules](http://im.snibgo.com/customim.htm), with instructions on building IM with my modules.

If you use bash, it's worth looking at Fred's (@fmw42) scripts. For example, a histogram-matching script at http://www.fmwconcepts.com/imagemagick/histmatch/index.php .

> Ultimately, say I'm able to figure out the histogram matching, would I be able to apply the change to multiple images which are similar, or would the shift only work for identical images?

Yes, if you want. When you have a clut that transforms an image to make its histogram match the histogram of another image, the same clut can then be used on other images to get similar transformations.

**Reply by memphis-10** [reply]
Thank you.  I think CLUT is a good way to go.  I think I can make WB and exposure adjustments in an external editor then apply.

**Reply by memphis-10** [reply]
> "call" is a Windows command that executes a .BAT script file. I publish these on my website im.snibgo.com . Most of my scripts can be easily translated to bash or any other shell. However, `matchHisto.bat` calls a process module. These are like plug-ins for IM, written in C, and compiled into the IM build. I publish these, mostly at [Process modules](http://im.snibgo.com/customim.htm), with instructions on building IM with my modules.

Thank you.  I have studied your scripts, but am still learning how I might implement them in a one-off IM command, or modify portions for specific use.  I don't often have this need, however, and my scripting knowledge is limited.  Building custom IM is certainly interesting and I hope to look into that further.

> Yes, if you want. When you have a clut that transforms an image to make its histogram match the histogram of another image, the same clut can then be used on other images to get similar transformations.

I created a CLUT in Lightroom that mostly does the trick---raises the exposure and saves the highlights (the Canon app only works on their raw files).  Interestingly, the Lightroom export and IM CLUT transform aren't identical, but they are extremely close, and an excellent starting point for the next steps (testing on similar images).

**Reply by memphis-10** [reply]
Hi @snibgo in a related question, part of my overall processing project, but not pertaining to these examples, I'm attempting to use Fred's histomatch and matchimage scripts to match a flat, uncorrected image, to the same image that's had adjustments.

I'm able to get extremely close with his two scripts (both, in that order), but not quite exact.  He said it's likely a matter of IM's histogram tools being limited to 256 bins and to check with you.  Also, my master is 16bits where the adjusted file I want to match to is 8bits.

Do you have any scripts or suggestions - also is there a better place to ask this question?  thank you

**Reply by fmw42** [reply]
I am confused. Which of your two images are you trying to match to which?  The 16-bit to the 8-bit or the 8-bit to the 16-bit? Is your 16-bit image actually using all 16-bits or are there regular gaps as in an 8-bit image stretched to 16-bits?

IM's histogram could have more bins, but likely 65535 for a 16-bit image would not be practical. But I am not sure it is controllable from the command line.  I just have not used them as I convert my images to 8-bits to make the number of bins practical for a script to process.

**Reply by memphis-10** [reply]
Sorry for any confusion.  I just replied to the thread in your repo.

I have a 16-bit file
It was edited and saved as 8-bit

I would like to reproduce the edits in the 8-bit file, but continue editing in 16-bit

**Reply by fmw42** [reply]
So do you know what you did to process the file? If so, repeat but do not save to 8-bits.

**Reply by memphis-10** [reply]
Thank you.  Therein lies the main issue: no, I do not have access to the original processing information.

**Comment by fmw42** [comment]
Doing a test on Q16 IM 7:

```
magick lena.png -format "%c" histogram:info: | wc -l
   48183

```
So IM 7 is listing all (48183 unique) colors (up to quantumrange --- 65535) in the image. So more bins are available. But my script would only be able to use that number. I could not set it easily to combine bins so that has say 1024 bins.  I am not sure how much slower that would be. I would have to try that.

But 48183 is for the color image. My script processed by channel. For example the red channel only has 201 bins in its histogram.

**Reply by memphis-10** [reply]
Thank you for checking and offering your detailed findings.  Please do keep me posted.  Speed isn't a concern at the moment.

**Comment by snibgo** [comment]
On the number of buckets: yes, my modules are written in C. The number of buckets is usually capped to the bit depth of the image. For example, a 16-bit image will have 65536 buckets. A 32-float image can have more buckets, eg 1 million. (use "`mkhisto ...  capnumbuckets 1000000`".)

Always remember that matching histograms will do only that. It makes a CLUT that will transform one image so that its histograms roughly matches the histograms of another image. (There are usually three histograms, one per RGB channel.) It cannnot reproduce all possible colour edits. For example, when a red pixel has a value of 75%, an edit may lighten some of those R=75% values (eg we want to lighten green grass) and darken other of those R=75% values (eg we want to darken skin tones). The histogram matching can't do both. It can either lighten R=75% or darken R=75%.

In principle, A 3-D histogram could be made, that counted the nuber of pixels that fall into buckets indexed by all three colour channels. However, that histogram will be sparse, so the corresponding Hald-clut will be sparse.

In a typical photograph, most pixels will fall in the middle half of the buckets. So a histogram of 256 buckets only has about 128 relevant buckets, so a loss of precision may be noticeable.

@memphis-10: If you upload sample images somewhere, I can easily put them through my scripts. Then you can see whether the extra depth is worthwhile.

**Reply by memphis-10** [reply]
Thank you.  I'll work on an set of images for upload to you both.

fmw42's histogram matching script got me close, then running the resulting file through his image matching script (comparing to the same starting file) got it nearly perfect.  I believe the latter is the mean and SD method.  If this is as close as I can get, that's good to know.

It was only in this thread that I realized the scripts were also making 8-bit files.  My goal is to do my best to apply the 8-bit changes to my 16-bit file.

I don't mind if the gaps in the new histogram are filled with estimations/averages so long as there's information there.  if that makes sense.

**Reply by fmw42** [reply]
Your link says "Item Deleted".  There are not images there.

**Reply by memphis-10** [reply]
I deleted the link because I want to keep the master image private, but if you need it again, I'm happy to repost.

**Comment by snibgo** [comment]
> I have a 16-bit file
It was edited and saved as 8-bit

So you have a 16 bits/channel/pixel image that has been edited and saved as 8 bits/channel/pixel.

Then the output file has a maximum of 256 values per channel. A histogram with more than 256 buckets won't give greater precision because the data doesn't support it. But the clut can still be used on a 16-bit file, of course.

My match-histogram process gives this:
```
call %PICTBAT%matchHisto tree_16bit_flat.tif tree_8bit_master.tif

%IMG7%magick tree_16bit_flat_mh.tif -resize 600 tree_16bit_flat_mh.jpg
```
![tree_16bit_flat_mh](https://user-images.githubusercontent.com/33812368/188294994-5e41a95a-508c-49e3-9167-f3a997046c57.jpg)

This makes an output that is RMSE 2.07% from the desired tree_8bit_master.tif. The differences are especially obvious in the high-saturation red, but also in the green foliage. If the edit applied a transformation that could be accurately modelled by an 8-bit clut, I would expect an RMSE around 0.5% or better.

We can graph the transformation clut, tree_16bit_flat_mhcl_glc.png:
![tree_16bit_flat_mhcl_glc](https://user-images.githubusercontent.com/33812368/188294974-a0b330a9-d757-4cec-a113-00a09036578d.png)

**Reply by memphis-10** [reply]
Thank you, @snibgo.  Your result is similar to what I was getting with Fred's histogram matching script.  

So I understand the matched histogram won't have more than 256 colors/channel.  However, by applying the CLUT to a 16bit image, does that create the opportunity for greater precision when editing?  Would this allow for the additional colors to emerge, so-to-speak?  (i.e. if I were to make adjustments to to reds, shadows, contrast, brightness, etc. in the new 16bit image, would it provide greater accuracy than the 8-bit)?  Or is that akin to converting a JPG to TIF hoping for greater accuracy (once it's gone, it's never coming back)?

**Comment by snibgo** [comment]
> once it's gone, it's never coming back

Exactly.

Instead of colours and pixel values, suppose we consider two units of length: kilometres and miles.

I give you a true fact: 16.6789321 kilometres is 10 miles, to the nearest mile.

I have quoted the km with 9 significant digits, but miles to 2 only significant digits. "To the nearest mile" means the true distance is between 9.5 miles and 10.5 miles.

Now I ask you to convert 166.789321 kilometres to miles. How many miles is that? You spot that I have moved the decimal point by one place, so you quickly reply, "100 miles."

That answer is correct within a certain precision, which is plus or minus one part in 20. So better answers would be, "It is 100 miles, to the nearest 10 miles" or "It is between 95 and 105 miles."

A wrong answer would be "100 miles, to the nearest mile." That would over-state the precision.

**Reply by memphis-10** [reply]
Thank you.  I believe I understand.

If I match the histogram of a 16bit image to that of an 8bit image, the 16bit image will only have 256 colors/channel to work with?

If that's true, it begs the question, can I "stretch" a histogram from 8 to 16bits, using averages and interpolation from one bin to the next to fill gaps?

**Reply by memphis-10** [reply]
I think I already know the answer, but I had to ask.

**Comment by snibgo** [comment]
> If that's true, it begs the question, can I "stretch" a histogram from 8 to 16bits, using averages and interpolation from one bin to the next to fill gaps?

Certainly. That happens automagically. 

I uploaded the clut I created to http://im.snibgo.com/tree_16bit_flat_mhcl.tiff . It has 65536 buckets, but only 6691 unique colours. 256 will be "correct", and the others have been interpolated.

You can download and use that clut like this:
```
wget http://im.snibgo.com/tree_16bit_flat_mhcl.tiff

magick tree_16bit_flat.tif tree_16bit_flat_mhcl.tiff -clut x.png
```
... and x.png will look like the output I showed above.

The clut can be scaled or resized down, eg to 256 buckets, and it will still work, but the output will be slightly different. Any size of clut can be used with images of any depth; IM will interpolate as required. And we can do other operations, such as blurring the clut which will increase the number of unique colours and smooth out bumps in the graph of the clut. This won't increase the accuracy (we can't get back data we lost from the 8-bit image) but it might reduce artefacts such as banding.

**Reply by memphis-10** [reply]
Cheers, thank you. Even making progress with artifacts and banding will be helpful.

I appreciate the time and lessons.

> automagically

I missed that the fist time!

**Comment by fmw42** [comment]
Hello,

There is also a variation called matchimage.

Fred
